continuous_queries.md

```diff
@@ -0,0 +1,236 @@
+CREATE CONTINUOUS QUERY 1h_cpu_load
+ON database_name
+BEGIN
+  SELECT mean(value) as value, region
+  INTO 1h.cpu_load
+  FROM cpu_load
+  GROUP BY time(1h), region
+END
+```
+
+Or chain them together:
+
+```sql
+CREATE CONTINUOUS QUERY 10m_event_count
+ON database_name
+BEGIN
+  SELECT count(value)
+  INTO 10m.events
+  FROM events
+  GROUP BY time(10m)
+END
+
+-- this selects from the output of one continuous query and outputs to another series
+CREATE CONTINUOUS QUERY 1h_event_count
+ON database_name
+BEGIN
+  SELECT sum(count) as count
+  INTO 1h.events
+  FROM events
+  GROUP BY time(1h)
+END
+```
+
+Or multiple aggregations from all series in a measurement. This example assumes you have a retention policy named `1h`.
+
+```sql
+CREATE CONTINUOUS QUERY 1h_cpu_load
+ON database_name
+BEGIN
+  SELECT mean(value), percentile(80, value) as percentile_80, percentile(95, value) as percentile_95
+  INTO 1h.cpu_load
+  FROM cpu_load
+  GROUP BY time(1h), *
+END
+```
+
+The `GROUP BY *` indicates that we want to group by the tagset of the points written in. The same tags will be written to the output series. The multiple aggregates in the `SELECT` clause (percentile, mean) will be written in as fields to the resulting series.
+
+Showing what continuous queries we have:
+
+```sql
+LIST CONTINUOUS QUERIES
+```
+
+Dropping continuous queries:
+
+```sql
+DROP CONTINUOUS QUERY <name>
+ON <database>
+```
+
+### Security
+
+To create or drop a continuous query, the user must be an admin.
+
+### Limitations
+
+In order to prevent cycles and endless copying of data, the following limitation is enforced on continuous queries at create time:
+
+*The output of a continuous query must go to either a different measurement or to a different retention policy.*
+
+In theory they'd still be able to create a cycle with multiple continuous queries. We should check for these and disallow.
+
+## Proposed Architecture
+
+Continuous queries should be stored in the metastore cluster wide. That is, they amount to database schema that should be stored in every server in a cluster.
+
+Continuous queries will have to be handled in a different way for two different use cases: those that simply copy data (CQs without a group by time) and those that aggregate and downsample data (those with a group by time).
+
+### No group by time
+
+For CQs that have no `GROUP BY time` clause, they should be evaluated at the data node as part of the write. The single write should create any other writes for the CQ and submit those in the same request to the brokers to ensure that all writes succeed (both the original and the new CQ writes) or none do.
+
+I imagine the process going something like this:
+
+1. Convert the data point into its compact form `<series id><time><values>`
+2. For each CQ on the measurement and retention policy without a group by time:
+3. Run the data point through a special query engine that will output 0 or 1 data point
+4. Goto #1 for each newly generated data point
```

Comment:
It sounds like you're trying to deal with a CQ on CQ-generated data -- a chain. Is that what you're being careful to cover here?